mitodl/mit-open
# Addresses performance issues for the topics API

### What are the relevant tickets?




Closes https://github.com/mitodl/hq/issues/4828

### Description (What does it do?)

This addresses a performance issue created by N+1 query issues due to a hardcoded query in the serializer. I took the approach of leaving in the existing code as a fallback but otherwise annotating the query for better performance. I also put in place a query count assertion on testing this API

I saw performance improve at least an order of magnitude and it'll probably run better in deployed environments:

Before:
![Screenshot 2024-07-10 at 16-23-21 Silky - Requests](https://github.com/mitodl/mit-open/assets/28598/f1aa45ee-456c-4c32-a844-b6dde932a719)
After:
![Screenshot 2024-07-11 at 13-08-36 Silky - Requests](https://github.com/mitodl/mit-open/assets/28598/bfc9f3af-118b-43a5-bc4d-2925add6ad33)


### How can this be tested?

- Hit the `/api/v1/topics/` api on the `main` branch and observe that it's slow (seconds).
- Hit the same API on this branch and it should be much faster (milliseconds).
- You should be able to go to `/silk` and it requires you to be a logged in superuser to access.
